sonar-batch-protocol/src/main/java/org/sonar/batch/protocol/viewer/TextLineNumber.java

```diff
@@ -0,0 +1,460 @@
+    int lineHeight = fontMetrics.getHeight();
+    int y = r.y + r.height;
+    int descent = 0;
+
+    // The text needs to be positioned above the bottom of the bounding
+    // rectangle based on the descent of the font(s) contained on the row.
+
+    if (r.height == lineHeight) // default font is being used
+    {
+      descent = fontMetrics.getDescent();
+    } else // We need to check all the attributes for font changes
+    {
+      if (fonts == null)
+        fonts = new HashMap<String, FontMetrics>();
+
+      Element root = component.getDocument().getDefaultRootElement();
+      int index = root.getElementIndex(rowStartOffset);
+      Element line = root.getElement(index);
+
+      for (int i = 0; i < line.getElementCount(); i++) {
+        Element child = line.getElement(i);
+        AttributeSet as = child.getAttributes();
+        String fontFamily = (String) as.getAttribute(StyleConstants.FontFamily);
+        Integer fontSize = (Integer) as.getAttribute(StyleConstants.FontSize);
+        String key = fontFamily + fontSize;
+
+        FontMetrics fm = fonts.get(key);
+
+        if (fm == null) {
+          Font font = new Font(fontFamily, Font.PLAIN, fontSize);
+          fm = component.getFontMetrics(font);
+          fonts.put(key, fm);
+        }
+
+        descent = Math.max(descent, fm.getDescent());
+      }
+    }
+
+    return y - descent;
+  }
+
+  //
+  // Implement CaretListener interface
+  //
+  @Override
+  public void caretUpdate(CaretEvent e) {
+    // Get the line the caret is positioned on
+
+    int caretPosition = component.getCaretPosition();
+    Element root = component.getDocument().getDefaultRootElement();
+    int currentLine = root.getElementIndex(caretPosition);
+
+    // Need to repaint so the correct line number can be highlighted
+
+    if (lastLine != currentLine) {
+      repaint();
+      lastLine = currentLine;
+    }
+  }
+
+  //
+  // Implement DocumentListener interface
+  //
+  @Override
+  public void changedUpdate(DocumentEvent e) {
+    documentChanged();
+  }
+
+  @Override
+  public void insertUpdate(DocumentEvent e) {
+    documentChanged();
+  }
+
+  @Override
+  public void removeUpdate(DocumentEvent e) {
+    documentChanged();
+  }
+
+  /*
+   * A document change may affect the number of displayed lines of text.
+   * Therefore the lines numbers will also change.
+   */
+  private void documentChanged() {
+    // View of the component has not been updated at the time
+    // the DocumentEvent is fired
+
+    SwingUtilities.invokeLater(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          int endPos = component.getDocument().getLength();
+          Rectangle rect = component.modelToView(endPos);
+
+          if (rect != null && rect.y != lastHeight) {
+            setPreferredWidth();
+            repaint();
+            lastHeight = rect.y;
+          }
+        } catch (BadLocationException ex) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1166)